baybe/surrogates/gaussian_process/core.py

```diff
@@ -102,26 +143,26 @@ def _fit(self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor) -> No
 
         # define the covariance module for the numeric dimensions
         base_covar_module = self.kernel_factory(
-            searchspace, train_x, train_y
+            context.searchspace, train_x, train_y
         ).to_gpytorch(
-            ard_num_dims=train_x.shape[-1] - n_task_params,
-            active_dims=numeric_idxs,
+            ard_num_dims=train_x.shape[-1] - context.n_task_dimensions,
+            active_dims=numerical_idxs,
             batch_shape=batch_shape,
         )
 
         # create GP covariance
-        if task_idx is None:
+        if not context.is_multitask:
             covar_module = base_covar_module
         else:
             task_covar_module = gpytorch.kernels.IndexKernel(
-                num_tasks=searchspace.n_tasks,
-                active_dims=task_idx,
-                rank=searchspace.n_tasks,  # TODO: make controllable
+                num_tasks=context.n_tasks,
+                active_dims=context.task_idx,
+                rank=context.n_tasks,  # TODO: make controllable
```

Comment:
i know this comment was already present before PR, but I wonder what does it mean and why would it be something else than `context.n_tasks` ?

Reply:
It means we should let the user control this, i.e. probably by turning it into an attribute with a sane default. But I suggest to do this at the very end, when we also make the switch to the built-in `MultiTaskGP` class (which is possible now that they implemented the interface extension I asked for).

For the second question: this is a perfect example where I could see your magic square root rule come into play 😄 The rank controls the ... well ... rank of the correlation matrix to be learned, i.e. it dictates the number of free parameters to be determined from data. If I'm not mistaken, with their low-rank approximation, setting it to `n_tasks` would mean that you need to learn `n_tasks^2` parameters, so you can easily overfit (in my expectation, the `sqrt` would exactly cancel the square). Will verify these mechanics before we change the default here. In our existing use cases, this was not super relevant, since the number of source tasks was mostly `N=1` or some small number.

Reply:
makes sense many thx for the details